From 1f032498a111ae63e39a66149854e1809857cb82 Mon Sep 17 00:00:00 2001
From: Sammie Quon <sammiequon@chromium.org>
Date: Tue, 19 Jul 2022 13:30:23 +0000
Subject: [PATCH 108/110] [M102-LTS] overview: Fix u-a-f when snapping a
 fullscreen window in tablet.

Fixed: 1330042
Test: ash_unittests TabletModeOverviewSessionTest.SnappingFullscreenWindow

(cherry picked from commit cb8d2046dccda41a687b9abe537a2d7361572c57)

Change-Id: I41ae7af0254ca69dc5bfa06bd860ffa62c2c7bef
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3692092
Commit-Queue: Sammie Quon <sammiequon@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1012093}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3755551
Commit-Queue: Roger Felipe Zanoni da Silva <rzanoni@google.com>
Owners-Override: Simon Hangl <simonha@google.com>
Reviewed-by: Simon Hangl <simonha@google.com>
Cr-Commit-Position: refs/branch-heads/5005@{#1267}
Cr-Branched-From: 5b4d9450fee01f821b6400e947b3839727643a71-refs/heads/main@{#992738}
---
 ash/wm/overview/overview_session.cc           |  3 ++-
 ash/wm/overview/overview_session_unittest.cc  | 25 +++++++++++++++++++
 .../overview_window_drag_controller.cc        |  9 +++++--
 3 files changed, 34 insertions(+), 3 deletions(-)

diff --git a/ash/wm/overview/overview_session.cc b/ash/wm/overview/overview_session.cc
index b5cee1d6d68f..aa8faf3b1a69 100644
--- a/ash/wm/overview/overview_session.cc
+++ b/ash/wm/overview/overview_session.cc
@@ -1157,7 +1157,8 @@ void OverviewSession::OnDisplayAdded(const display::Display& display) {
 
 void OverviewSession::OnDisplayMetricsChanged(const display::Display& display,
                                               uint32_t metrics) {
-  if (window_drag_controller_)
+  // End the current drag if the display changes.
+  if (window_drag_controller_ && window_drag_controller_->item())
     ResetDraggedWindowGesture();
   auto* overview_grid =
       GetGridWithRootWindow(Shell::GetRootWindowForDisplayId(display.id()));
diff --git a/ash/wm/overview/overview_session_unittest.cc b/ash/wm/overview/overview_session_unittest.cc
index a19be47a0ba9..50e3bee61c12 100644
--- a/ash/wm/overview/overview_session_unittest.cc
+++ b/ash/wm/overview/overview_session_unittest.cc
@@ -3677,6 +3677,31 @@ TEST_F(TabletModeOverviewSessionTest, HorizontalScrollingOnOverviewItem) {
   EXPECT_LT(leftmost_window->target_bounds(), left_bounds);
 }
 
+// Tests that dragging a fullscreened window to snap in overview does not result
+// in a u-a-f. Regression test for crbug.com/1330042.
+TEST_F(TabletModeOverviewSessionTest, SnappingFullscreenWindow) {
+  UpdateDisplay("800x600");
+
+  auto window = CreateAppWindow(gfx::Rect(300, 300));
+
+  const WMEvent fullscreen_event(WM_EVENT_FULLSCREEN);
+  WindowState::Get(window.get())->OnWMEvent(&fullscreen_event);
+  EXPECT_TRUE(WindowState::Get(window.get())->IsFullscreen());
+
+  ToggleOverview();
+  ASSERT_TRUE(InOverviewSession());
+
+  OverviewItem* item = GetOverviewItemForWindow(window.get());
+  ui::test::EventGenerator* generator = GetEventGenerator();
+  generator->set_current_screen_location(
+      gfx::ToRoundedPoint(item->target_bounds().CenterPoint()));
+  generator->PressLeftButton();
+  generator->MoveMouseTo(gfx::Point(10, 300));
+  generator->ReleaseLeftButton();
+
+  EXPECT_TRUE(WindowState::Get(window.get())->IsSnapped());
+}
+
 // A unique test class for testing flings in overview as those rely on observing
 // compositor animations which require a mock time task environment.
 class OverviewSessionFlingTest : public AshTestBase {
diff --git a/ash/wm/overview/overview_window_drag_controller.cc b/ash/wm/overview/overview_window_drag_controller.cc
index 3d0f80aa73cb..c61b317c8f9e 100644
--- a/ash/wm/overview/overview_window_drag_controller.cc
+++ b/ash/wm/overview/overview_window_drag_controller.cc
@@ -811,16 +811,21 @@ void OverviewWindowDragController::SnapWindow(
     SplitViewController::SnapPosition snap_position) {
   DCHECK_NE(snap_position, SplitViewController::NONE);
 
-  // |item_| will be deleted after SplitViewController::SnapWindow().
   DCHECK(!SplitViewController::Get(Shell::GetPrimaryRootWindow())
               ->IsDividerAnimating());
   aura::Window* window = item_->GetWindow();
   WindowState::Get(window)->set_snap_action_source(
       WindowSnapActionSource::kDragOrSelectOverviewWindowToSnap);
 
+  // If `window` is currently fullscreen, snapping it will trigger a work area
+  // change, which triggers `OverviewSession::OnDisplayMetricsChanged`. Display
+  // changes normally end dragging for simplicity, but we need `item` to be
+  // nullptr before that happens so we can skip resetting the window gesture.
+  // See crbug.com/1330042 for more details. `item_` will be deleted after
+  // SplitViewController::SnapWindow().
+  item_ = nullptr;
   split_view_controller->SnapWindow(window, snap_position,
                                     /*activate_window=*/true);
-  item_ = nullptr;
 }
 
 OverviewGrid* OverviewWindowDragController::GetCurrentGrid() const {
-- 
2.38.1

